### ISSUE TITLE
Crashes: assertions fail.

### ISSUE THREAD

**Issue by mcourteaux** [wolfpld/tracy#875]
I updated Tracy after a really long time, and now I'm getting consistently crashes. Running a debug-build in gdb gives me:

> tracy-profiler: /home/martijn/zec/3rd/tracy/profiler/src/profiler/TracyTimelineItemThread.cpp:289: virtual void tracy::TimelineItemThread::Preprocess(const tracy::TimelineContext&, tracy::TaskDispatch&, bool, int): Assertion `m_ctxDraw.empty()' failed.

```cpp
(gdb) bt
#0  __pthread_kill_implementation (no_tid=0, signo=6, threadid=140737345933696) at /home/martijn/zec/3rd/tracy/profiler/build_dbg/nptl/pthread_kill.c:44
#1  __pthread_kill_internal (signo=6, threadid=140737345933696) at /home/martijn/zec/3rd/tracy/profiler/build_dbg/nptl/pthread_kill.c:78
#2  __GI___pthread_kill (threadid=140737345933696, signo=6) at /home/martijn/zec/3rd/tracy/profiler/build_dbg/nptl/pthread_kill.c:89
#3  0x00007ffff7642476 in __GI_raise (sig=6) at /home/martijn/zec/3rd/tracy/profiler/sysdeps/posix/raise.c:26
#4  0x00007ffff76287f3 in __GI_abort () at /home/martijn/zec/3rd/tracy/profiler/build_dbg/stdlib/abort.c:79
#5  0x00007ffff762871b in __assert_fail_base (fmt=0x7ffff77dd130 "%s%s%s:%u: %s%sAssertion `%s' failed.\n%n", assertion=0x555556dfe3bc "m_ctxDraw.empty()", file=0x555556dfe358 "/home/martijn/zec/3rd/tracy/profiler/src/profiler/TracyTimelineItemThread.cpp", line=289, function=<optimized out>) at /home/martijn/zec/3rd/tracy/profiler/build_dbg/assert/assert.c:92
#6  0x00007ffff7639e96 in __GI___assert_fail (assertion=0x555556dfe3bc "m_ctxDraw.empty()", file=0x555556dfe358 "/home/martijn/zec/3rd/tracy/profiler/src/profiler/TracyTimelineItemThread.cpp", line=289, function=0x555556dfe2e0 "virtual void tracy::TimelineItemThread::Preprocess(const tracy::TimelineContext&, tracy::TaskDispatch&, bool, int)")
    at /home/martijn/zec/3rd/tracy/profiler/build_dbg/assert/assert.c:101
#7  0x0000555555ca03a5 in tracy::TimelineItemThread::Preprocess(tracy::TimelineContext const&, tracy::TaskDispatch&, bool, int) (this=0x55555824dc40, ctx=..., td=..., visible=true, yPos=100) at /home/martijn/zec/3rd/tracy/profiler/src/profiler/TracyTimelineItemThread.cpp:289
#8  0x0000555555c6f988 in tracy::TimelineController::End(double, ImVec2 const&, bool, bool, float, float, ImFont*) (this=0x55555853ae70, pxns=7.776086254290375e-06, wpos=..., hover=false, vcenter=false, yMin=128, yMax=548, smallFont=0x555557e02ed0) at /home/martijn/zec/3rd/tracy/profiler/src/profiler/TracyTimelineController.cpp:146
#9  0x0000555555efb46c in tracy::View::DrawTimeline() (this=0x555558538900) at /home/martijn/zec/3rd/tracy/profiler/src/profiler/TracyView_Timeline.cpp:389
#10 0x0000555555cf0a4d in tracy::View::DrawImpl() (this=0x555558538900) at /home/martijn/zec/3rd/tracy/profiler/src/profiler/TracyView.cpp:1097
#11 0x0000555555cedfd3 in tracy::View::Draw() (this=0x555558538900) at /home/martijn/zec/3rd/tracy/profiler/src/profiler/TracyView.cpp:654
#12 0x0000555555bb5ec5 in DrawContents() () at /home/martijn/zec/3rd/tracy/profiler/src/main.cpp:1190
#17 0x0000555555bcecd2 in Backend::Run() (this=0x7fffffffdad0) at /home/martijn/zec/3rd/tracy/profiler/src/BackendGlfw.cpp:152
#18 0x0000555555bb1d79 in main(int, char**) (argc=1, argv=0x7fffffffe018) at /home/martijn/zec/3rd/tracy/profiler/src/main.cpp:418

(gdb) f 7
#7  0x0000555555ca03a5 in tracy::TimelineItemThread::Preprocess (this=0x55555824dc40, ctx=..., td=..., visible=true, yPos=100) at /home/martijn/zec/3rd/tracy/profiler/src/profiler/TracyTimelineItemThread.cpp:289
289	   assert( m_ctxDraw.empty() );
(gdb) p m_ctxDraw 
$1 = std::vector of length 83, capacity 128 = {{type = tracy::ContextSwitchDrawType::Folded, idx = 979, data = 1}, {type = tracy::ContextSwitchDrawType::Waiting, idx = 980, data = 23156}, {type = tracy::ContextSwitchDrawType::Folded, idx = 980, data = 1}, {type = tracy::ContextSwitchDrawType::Waiting, idx = 981, data = 23156}, {type = tracy::ContextSwitchDrawType::Folded, 
    idx = 981, data = 1}, {type = tracy::ContextSwitchDrawType::Waiting, idx = 982, data = 23156}, {type = tracy::ContextSwitchDrawType::Folded, idx = 982, data = 1}, {type = tracy::ContextSwitchDrawType::Waiting, idx = 983, data = 23156}, {type = tracy::ContextSwitchDrawType::Folded, idx = 983, data = 1}, {type = tracy::ContextSwitchDrawType::Waiting, idx = 984, data = 23156}, 
  {type = tracy::ContextSwitchDrawType::Folded, idx = 984, data = 1}, {type = tracy::ContextSwitchDrawType::Waiting, idx = 985, data = 23156}, {type = tracy::ContextSwitchDrawType::Folded, idx = 985, data = 2}, {type = tracy::ContextSwitchDrawType::Waiting, idx = 987, data = 23156}, {type = tracy::ContextSwitchDrawType::Folded, idx = 987, data = 1}, {
    type = tracy::ContextSwitchDrawType::Waiting, idx = 988, data = 23156}, {type = tracy::ContextSwitchDrawType::Folded, idx = 988, data = 1}, {type = tracy::ContextSwitchDrawType::Waiting, idx = 989, data = 23156}, {type = tracy::ContextSwitchDrawType::Folded, idx = 989, data = 1}, {type = tracy::ContextSwitchDrawType::Waiting, idx = 990, data = 23156}, {
    type = tracy::ContextSwitchDrawType::Folded, idx = 990, data = 1}, {type = tracy::ContextSwitchDrawType::Waiting, idx = 991, data = 23156}, {type = tracy::ContextSwitchDrawType::Folded, idx = 991, data = 1}, {type = tracy::ContextSwitchDrawType::Waiting, idx = 992, data = 23156}, {type = tracy::ContextSwitchDrawType::Folded, idx = 992, data = 1}, {
    type = tracy::ContextSwitchDrawType::Waiting, idx = 993, data = 23156}, {type = tracy::ContextSwitchDrawType::Folded, idx = 993, data = 1}, {type = tracy::ContextSwitchDrawType::Waiting, idx = 994, data = 23156}, {type = tracy::ContextSwitchDrawType::Folded, idx = 994, data = 1}, {type = tracy::ContextSwitchDrawType::Waiting, idx = 995, data = 23156}, {
    type = tracy::ContextSwitchDrawType::Folded, idx = 995, data = 1}, {type = tracy::ContextSwitchDrawType::Waiting, idx = 996, data = 23156}, {type = tracy::ContextSwitchDrawType::Folded, idx = 996, data = 1}, {type = tracy::ContextSwitchDrawType::Waiting, idx = 997, data = 23156}, {type = tracy::ContextSwitchDrawType::Folded, idx = 997, data = 1}, {
    type = tracy::ContextSwitchDrawType::Waiting, idx = 998, data = 23156}, {type = tracy::ContextSwitchDrawType::Folded, idx = 998, data = 1}, {type = tracy::ContextSwitchDrawType::Waiting, idx = 999, data = 23156}, {type = tracy::ContextSwitchDrawType::Folded, idx = 999, data = 1}, {type = tracy::ContextSwitchDrawType::Waiting, idx = 1000, data = 23156}, {
    type = tracy::ContextSwitchDrawType::Folded, idx = 1000, data = 1}, {type = tracy::ContextSwitchDrawType::Waiting, idx = 1001, data = 23156}, {type = tracy::ContextSwitchDrawType::Folded, idx = 1001, data = 1}, {type = tracy::ContextSwitchDrawType::Waiting, idx = 1002, data = 23156}, {type = tracy::ContextSwitchDrawType::Folded, idx = 1002, data = 1}, {
    type = tracy::ContextSwitchDrawType::Waiting, idx = 1003, data = 23156}, {type = tracy::ContextSwitchDrawType::Folded, idx = 1003, data = 1}, {type = tracy::ContextSwitchDrawType::Waiting, idx = 1004, data = 23156}, {type = tracy::ContextSwitchDrawType::Folded, idx = 1004, data = 1}, {type = tracy::ContextSwitchDrawType::Waiting, idx = 1005, data = 23156}, {
    type = tracy::ContextSwitchDrawType::Folded, idx = 1005, data = 1}, {type = tracy::ContextSwitchDrawType::Waiting, idx = 1006, data = 23156}, {type = tracy::ContextSwitchDrawType::Folded, idx = 1006, data = 1}, {type = tracy::ContextSwitchDrawType::Waiting, idx = 1007, data = 23156}, {type = tracy::ContextSwitchDrawType::Folded, idx = 1007, data = 1}, {
    type = tracy::ContextSwitchDrawType::Waiting, idx = 1008, data = 23156}, {type = tracy::ContextSwitchDrawType::Folded, idx = 1008, data = 1}, {type = tracy::ContextSwitchDrawType::Waiting, idx = 1009, data = 23156}, {type = tracy::ContextSwitchDrawType::Folded, idx = 1009, data = 1}, {type = tracy::ContextSwitchDrawType::Waiting, idx = 1010, data = 23156}, {
    type = tracy::ContextSwitchDrawType::Folded, idx = 1010, data = 1}, {type = tracy::ContextSwitchDrawType::Waiting, idx = 1011, data = 23156}, {type = tracy::ContextSwitchDrawType::Folded, idx = 1011, data = 1}, {type = tracy::ContextSwitchDrawType::Waiting, idx = 1012, data = 23156}, {type = tracy::ContextSwitchDrawType::Folded, idx = 1012, data = 1}, {
    type = tracy::ContextSwitchDrawType::Waiting, idx = 1013, data = 23156}, {type = tracy::ContextSwitchDrawType::Folded, idx = 1013, data = 1}, {type = tracy::ContextSwitchDrawType::Waiting, idx = 1014, data = 23156}, {type = tracy::ContextSwitchDrawType::Folded, idx = 1014, data = 1}, {type = tracy::ContextSwitchDrawType::Waiting, idx = 1015, data = 23156}, {
    type = tracy::ContextSwitchDrawType::Folded, idx = 1015, data = 1}, {type = tracy::ContextSwitchDrawType::Waiting, idx = 1016, data = 23156}, {type = tracy::ContextSwitchDrawType::Folded, idx = 1016, data = 1}, {type = tracy::ContextSwitchDrawType::Waiting, idx = 1017, data = 23156}, {type = tracy::ContextSwitchDrawType::Folded, idx = 1017, data = 1}, {
    type = tracy::ContextSwitchDrawType::Waiting, idx = 1018, data = 23156}, {type = tracy::ContextSwitchDrawType::Folded, idx = 1018, data = 1}, {type = tracy::ContextSwitchDrawType::Waiting, idx = 1019, data = 23156}, {type = tracy::ContextSwitchDrawType::Folded, idx = 1019, data = 1}, {type = tracy::ContextSwitchDrawType::Waiting, idx = 1020, data = 23156}, {
    type = tracy::ContextSwitchDrawType::Folded, idx = 1020, data = 1}, {type = tracy::ContextSwitchDrawType::Waiting, idx = 1021, data = 23156}, {type = tracy::ContextSwitchDrawType::Folded, idx = 1021, data = 1}}
```

I haven't figured out the pattern yet.

---

Another run gave:

> tracy-profiler: /home/martijn/zec/3rd/tracy/profiler/src/profiler/TracyTimelineItemThread.cpp:290: virtual void tracy::TimelineItemThread::Preprocess(const tracy::TimelineContext&, tracy::TaskDispatch&, bool, int): Assertion `m_draw.empty()' failed.

```cpp
(gdb) bt
#0  __pthread_kill_implementation (no_tid=0, signo=6, threadid=140737345933696) at /home/martijn/zec/3rd/tracy/profiler/build_dbg/nptl/pthread_kill.c:44
#1  __pthread_kill_internal (signo=6, threadid=140737345933696) at /home/martijn/zec/3rd/tracy/profiler/build_dbg/nptl/pthread_kill.c:78
#2  __GI___pthread_kill (threadid=140737345933696, signo=6) at /home/martijn/zec/3rd/tracy/profiler/build_dbg/nptl/pthread_kill.c:89
#3  0x00007ffff7642476 in __GI_raise (sig=6) at /home/martijn/zec/3rd/tracy/profiler/sysdeps/posix/raise.c:26
#4  0x00007ffff76287f3 in __GI_abort () at /home/martijn/zec/3rd/tracy/profiler/build_dbg/stdlib/abort.c:79
#5  0x00007ffff762871b in __assert_fail_base (fmt=0x7ffff77dd130 "%s%s%s:%u: %s%sAssertion `%s' failed.\n%n", assertion=0x555556dfe3ce "m_draw.empty()", file=0x555556dfe358 "/home/martijn/zec/3rd/tracy/profiler/src/profiler/TracyTimelineItemThread.cpp", line=290, function=<optimized out>) at /home/martijn/zec/3rd/tracy/profiler/build_dbg/assert/assert.c:92
#6  0x00007ffff7639e96 in __GI___assert_fail (assertion=0x555556dfe3ce "m_draw.empty()", file=0x555556dfe358 "/home/martijn/zec/3rd/tracy/profiler/src/profiler/TracyTimelineItemThread.cpp", line=290, function=0x555556dfe2e0 "virtual void tracy::TimelineItemThread::Preprocess(const tracy::TimelineContext&, tracy::TaskDispatch&, bool, int)")
    at /home/martijn/zec/3rd/tracy/profiler/build_dbg/assert/assert.c:101
#7  0x0000555555ca03e4 in tracy::TimelineItemThread::Preprocess(tracy::TimelineContext const&, tracy::TaskDispatch&, bool, int) (this=0x5555580da350, ctx=..., td=..., visible=true, yPos=128) at /home/martijn/zec/3rd/tracy/profiler/src/profiler/TracyTimelineItemThread.cpp:290
#8  0x0000555555c6f988 in tracy::TimelineController::End(double, ImVec2 const&, bool, bool, float, float, ImFont*) (this=0x555558104cd0, pxns=1.9609920752517868e-05, wpos=..., hover=false, vcenter=false, yMin=128, yMax=308, smallFont=0x555557e02ed0) at /home/martijn/zec/3rd/tracy/profiler/src/profiler/TracyTimelineController.cpp:146
#9  0x0000555555efb46c in tracy::View::DrawTimeline() (this=0x555558102760) at /home/martijn/zec/3rd/tracy/profiler/src/profiler/TracyView_Timeline.cpp:389
#10 0x0000555555cf0a4d in tracy::View::DrawImpl() (this=0x555558102760) at /home/martijn/zec/3rd/tracy/profiler/src/profiler/TracyView.cpp:1097
#11 0x0000555555cedfd3 in tracy::View::Draw() (this=0x555558102760) at /home/martijn/zec/3rd/tracy/profiler/src/profiler/TracyView.cpp:654
#12 0x0000555555bb5ec5 in DrawContents() () at /home/martijn/zec/3rd/tracy/profiler/src/main.cpp:1190
#17 0x0000555555bcecd2 in Backend::Run() (this=0x7fffffffdad0) at /home/martijn/zec/3rd/tracy/profiler/src/BackendGlfw.cpp:152
#18 0x0000555555bb1d79 in main(int, char**) (argc=1, argv=0x7fffffffe018) at /home/martijn/zec/3rd/tracy/profiler/src/main.cpp:418
(gdb) f 7
#7  0x0000555555ca03e4 in tracy::TimelineItemThread::Preprocess (this=0x5555580da350, ctx=..., td=..., visible=true, yPos=128) at /home/martijn/zec/3rd/tracy/profiler/src/profiler/TracyTimelineItemThread.cpp:290
290	   assert( m_draw.empty() );
(gdb) p m_draw
$1 = std::vector of length 3, capacity 8 = {{type = tracy::TimelineDrawType::Folded, depth = 0, ev = {m_ptr = "\347Z\203\340\377\177"}, rend = {m_val = "0>1\246\000"}, num = 35}, {type = tracy::TimelineDrawType::Folded, depth = 0, ev = {m_ptr = "\351]\203\340\377\177"}, rend = {m_val = "\371E#\250\000"}, num = 31}, {type = tracy::TimelineDrawType::Folded, depth = 0, ev = {
      m_ptr = "\223`\203\340\377\177"}, rend = {m_val = "h\357\070\252\000"}, num = 35}}
(gdb) 
```

I was looking into the git history, but all files have moved, and I hope you know what's up, and I don't have to start git-bisect, because that will take a lot of time. I don't even know which commit I come from.

**Comment by wolfpld** [comment]
`TimelineItemThread::Preprocess()` builds a draw list. The assert you see checks if the previous list built for the object was cleared.

The cycle of build, draw, clear is entirely contained in `TimelineController::End()`. The `m_items` list is iterated and `TimelineItem::Preprocess()` is called on each item, which builds the draw list. Then `m_items` is iterated again and `TimelineItem::Draw()` is called on each item to draw the items. Within that function each exit path is calling `TimelineItem::DrawFinished()`, which in case of `TimelineItemThread` clears the `m_draw` list.

I don't see where this would be failing.

**Comment by mcourteaux** [comment]
> The cycle of build, draw, clear is entirely contained in `TimelineController::End()`.

I have no clue what the purpose of the code is, but it seems that `DrawFinished()` clears the lists. `DrawFinished()` is not called in `TimelineController::End()`.

However, my two crashes that I captured crash on different asserts one line apart. This feels like it could be a concurrency issue? The list gets cleared while the other thing is running asserts? I see there is an option for `NO_PARALLEL_STL`; any chance this is related? I might try to toy with that and see.

I tried recompiling the Tracy UI with AddressSanitizer, but now I can't catch the bug. It just doesn't appear anymore. Seems to be a Heisenbug.

I'll report if I find anything more.

**Comment by wolfpld** [comment]
> `DrawFinished()` is not called in `TimelineController::End()`.

It is called via `item->Draw()`.

> However, my two crashes that I captured crash on different asserts one line apart. This feels like it could be a concurrency issue?

The first `m_items` loop indeed puts the parallel tasks into `m_td` inside the `item->Preprocess()` call.

> The list gets cleared while the other thing is running asserts?

`Draw()` is only called from the second `m_items` loop (in a serial manner), and at this point all the parallel tasks should be finished (via the `m_td.Sync()` call).

> I see there is an option for `NO_PARALLEL_STL`; any chance this is related?

Unrelated. This only affects the STL function calls that have a `std::execution` policy parameter.

**Comment by mcourteaux** [comment]
> Within that function each exit path is calling `TimelineItem::DrawFinished()`, which in case of `TimelineItemThread` clears the `m_draw` list.
> 
> I don't see where this would be failing.

The issue seems to be that some items are multiple times in the list. I just confirmed this:

```
(gdb) p m_items
$1 = std::vector of length 28, capacity 32 = {0x61000000ad40, 0x61000000ae40, 0x61000000ae40, 0x61000000b040, 0x61000000b140, 0x61000000b240, 0x61000000b340, 0x61000000b440, 0x61000000b540, 0x61000000b640, 0x61000000b740, 0x60700004db80, 0x60700004dbf0, 0x60700004dc60, 0x60700004dcd0, 0x60700004dd40, 0x60700004de20, 0x60700004de90, 0x60700004df00, 0x60700004df70, 
  0x60700005a7d0, 0x60700005a840, 0x60700005a8b0, 0x60700005a920, 0x60700005a990, 0x60700005aa00, 0x60700005aa70, 0x60700005aae0}
(gdb) 
```

Notice how elements 2 and 3 are the same `0x61000000ae40`. We already know that those must be `TimelineItemThread` objects. Looking at where they get added, we see that's happening here:

https://github.com/wolfpld/tracy/blob/49f0ca53395ffa9e4dcc64567622be67be075691/profiler/src/profiler/TracyView_Timeline.cpp#L364-L379

Now the code above regarding thread-order was changed very recently, so maybe there something went wrong? So I do believe indeed that the problem is not really a race-condition, because the item is simply proven to be in the list twice.

Looking at the data in m_threadOrder:

```
(gdb) x/6ubx m_threadOrder.m_ptr.m_ptr 
0x629000064350:	0x40	0x56	0x07	0x00	0xc0	0x60
(gdb) p m_threadOrder
$11 = {m_ptr = {m_ptr = "@V\a\000\300`"}, m_size = 11, m_capacity = 4 '\004', m_magic = 0 '\000'}
(gdb) x/11a 0x60c000075640
0x60c000075640:	0x7fffcd4018d4	0x7fffcd4c1df3
0x60c000075650:	0x7fffcd4c1df3	0x7fffcd436f96
0x60c000075660:	0x7fffcd4474ef	0x7fffcd457a48
0x60c000075670:	0x7fffcd467fa1	0x7fffcd47807f
0x60c000075680:	0x7fffcd488a53	0x7fffcd498b9f
0x60c000075690:	0x7fffcd4a8d95
```

We indeed again see a repeated element in position 2 and 3 (counting from 1). Now looking at where that data comes from:

```
(gdb) x/6ubx m_worker.m_data.threads.m_ptr.m_ptr
0x629000064e70:	0x00	0x66	0x07	0x00	0xc0	0x60
(gdb) x/11a 0x60c000076600
0x60c000076600:	0x7fffcd4018d4	0x7fffcd426a3d
0x60c000076610:	0x7fffcd436f96	0x7fffcd4474ef
0x60c000076620:	0x7fffcd457a48	0x7fffcd467fa1
0x60c000076630:	0x7fffcd47807f	0x7fffcd488a53
0x60c000076640:	0x7fffcd498b9f	0x7fffcd4a8d95
0x60c000076650:	0x7fffcd4c1df3
```

I don't see any repeated elements. Additionally, it is clear that the last element (i.e. the last received thread data by the worker), matches the repeated element. So the bug is caught as soon as the thread data is repeated in the `m_threadOrder` vector. So clearly, something is wrong when copying the thread-data pointers over from the normal list, to the sorted list. Looking back at the code which does that, I don't see anything particularly wrong:

https://github.com/wolfpld/tracy/blob/49f0ca53395ffa9e4dcc64567622be67be075691/profiler/src/profiler/TracyView_Timeline.cpp#L369-L373

Regardless of what std::upper_bound exactly returns, this code looks good: never will `threadData[i]` be the same as before, since the `GetThreadData()` list is always grown by `push_back`. So this leaves my last suspect to be: `TracyVector::insert()`:

https://github.com/wolfpld/tracy/blob/49f0ca53395ffa9e4dcc64567622be67be075691/server/TracyVector.hpp#L164-L175

Asking for ChatGPT's opinion, only a few handful of **unrelated**-but-worth-mentioning issues are raised:
 - `m_ptr[dist] = v` is redundant over the placement-new from the line above.
 - `memmove()` is probably not a good idea on non-POD types for `T`. However, the placement-new suggests more C++-y stuff (meaning constructors and destructors probably should be allowed to do their thing).

I also don't see any problem in the memmove logic.

Different approach: printf debugging. Adding a print for inserting gives me:

```
[DEBUG this=0x62900005a200; m_threadOrder.size() =  0, worker.GetThreadData().size() = 11] Insert threadData[ 0]=0x7fffcd4018d4 in 0x60c000075640.
[DEBUG this=0x62900005a200; m_threadOrder.size() =  1, worker.GetThreadData().size() = 11] Insert threadData[ 1]=0x7fffcd426a3d in 0x60c000075648.
[DEBUG this=0x62900005a200; m_threadOrder.size() =  2, worker.GetThreadData().size() = 11] Insert threadData[ 2]=0x7fffcd436b1b in 0x60c000075650.
[DEBUG this=0x62900005a200; m_threadOrder.size() =  3, worker.GetThreadData().size() = 11] Insert threadData[ 3]=0x7fffcd447074 in 0x60c000075658.
[DEBUG this=0x62900005a200; m_threadOrder.size() =  4, worker.GetThreadData().size() = 11] Insert threadData[ 4]=0x7fffcd4575cd in 0x60c000075660.
[DEBUG this=0x62900005a200; m_threadOrder.size() =  5, worker.GetThreadData().size() = 11] Insert threadData[ 5]=0x7fffcd467b26 in 0x60c000075668.
[DEBUG this=0x62900005a200; m_threadOrder.size() =  6, worker.GetThreadData().size() = 11] Insert threadData[ 6]=0x7fffcd47807f in 0x60c000075670.
[DEBUG this=0x62900005a200; m_threadOrder.size() =  7, worker.GetThreadData().size() = 11] Insert threadData[ 7]=0x7fffcd4885d8 in 0x60c000075678.
[DEBUG this=0x62900005a200; m_threadOrder.size() =  8, worker.GetThreadData().size() = 11] Insert threadData[ 8]=0x7fffcd498b31 in 0x60c000075680.
[DEBUG this=0x62900005a200; m_threadOrder.size() =  9, worker.GetThreadData().size() = 11] Insert threadData[ 9]=0x7fffcd4a8c5f in 0x60c000075688.
[DEBUG this=0x62900005a200; m_threadOrder.size() = 10, worker.GetThreadData().size() = 11] Insert threadData[10]=0x7fffcd4bed25 in 0x60c000075648.
Done sorting.
[DEBUG this=0x62900005a200; m_threadOrder.size() =  2, worker.GetThreadData().size() = 11] Insert threadData[ 2]=0x7fffcd436b1b in 0x60c000075650.
[DEBUG this=0x62900005a200; m_threadOrder.size() =  3, worker.GetThreadData().size() = 11] Insert threadData[ 3]=0x7fffcd447074 in 0x60c000075658.
[DEBUG this=0x62900005a200; m_threadOrder.size() =  4, worker.GetThreadData().size() = 11] Insert threadData[ 4]=0x7fffcd4575cd in 0x60c000075660.
[DEBUG this=0x62900005a200; m_threadOrder.size() =  5, worker.GetThreadData().size() = 11] Insert threadData[ 5]=0x7fffcd467b26 in 0x60c000075668.
[DEBUG this=0x62900005a200; m_threadOrder.size() =  6, worker.GetThreadData().size() = 11] Insert threadData[ 6]=0x7fffcd47807f in 0x60c000075670.
[DEBUG this=0x62900005a200; m_threadOrder.size() =  7, worker.GetThreadData().size() = 11] Insert threadData[ 7]=0x7fffcd4885d8 in 0x60c000075678.
[DEBUG this=0x62900005a200; m_threadOrder.size() =  8, worker.GetThreadData().size() = 11] Insert threadData[ 8]=0x7fffcd498b31 in 0x60c000075680.
[DEBUG this=0x62900005a200; m_threadOrder.size() =  9, worker.GetThreadData().size() = 11] Insert threadData[ 9]=0x7fffcd4a8c5f in 0x60c000075688.
[DEBUG this=0x62900005a200; m_threadOrder.size() = 10, worker.GetThreadData().size() = 11] Insert threadData[10]=0x7fffcd4bed25 in 0x60c000075650.
Done sorting.
```

So somehow, the list gets shrunk down to 2 elements, and then regrown. Ripgrepping the codebase, I found inside `View::DrawImpl`:

https://github.com/wolfpld/tracy/blob/49f0ca53395ffa9e4dcc64567622be67be075691/profiler/src/profiler/TracyView.cpp#L715-L724

Adding a print for every `erase()` happening indeed uncovers this is the reason:

```
[DEBUG this=0x62900005a200; m_threadOrder.size() =  0, worker.GetThreadData().size() = 11] Insert threadData[ 0]=0x7fffccc018d4 in 0x60c000075640.
[DEBUG this=0x62900005a200; m_threadOrder.size() =  1, worker.GetThreadData().size() = 11] Insert threadData[ 1]=0x7fffccc26a3d in 0x60c000075648.
[DEBUG this=0x62900005a200; m_threadOrder.size() =  2, worker.GetThreadData().size() = 11] Insert threadData[ 2]=0x7fffccc36f96 in 0x60c000075650.
[DEBUG this=0x62900005a200; m_threadOrder.size() =  3, worker.GetThreadData().size() = 11] Insert threadData[ 3]=0x7fffccc47074 in 0x60c000075658.
[DEBUG this=0x62900005a200; m_threadOrder.size() =  4, worker.GetThreadData().size() = 11] Insert threadData[ 4]=0x7fffccc575cd in 0x60c000075660.
[DEBUG this=0x62900005a200; m_threadOrder.size() =  5, worker.GetThreadData().size() = 11] Insert threadData[ 5]=0x7fffccc67b26 in 0x60c000075668.
[DEBUG this=0x62900005a200; m_threadOrder.size() =  6, worker.GetThreadData().size() = 11] Insert threadData[ 6]=0x7fffccc784fa in 0x60c000075670.
[DEBUG this=0x62900005a200; m_threadOrder.size() =  7, worker.GetThreadData().size() = 11] Insert threadData[ 7]=0x7fffccc88a53 in 0x60c000075678.
[DEBUG this=0x62900005a200; m_threadOrder.size() =  8, worker.GetThreadData().size() = 11] Insert threadData[ 8]=0x7fffccc98b31 in 0x60c000075680.
[DEBUG this=0x62900005a200; m_threadOrder.size() =  9, worker.GetThreadData().size() = 11] Insert threadData[ 9]=0x7fffccca8cc3 in 0x60c000075688.
[DEBUG this=0x62900005a200; m_threadOrder.size() = 10, worker.GetThreadData().size() = 11] Insert threadData[10]=0x7fffcccbe3ad in 0x60c000075648.
Done sorting.
[DEBUG this=0x62900005a200] Erase 239060 from 0x60c000075650.
[DEBUG this=0x62900005a200] Erase 239059 from 0x60c000075650.
[DEBUG this=0x62900005a200] Erase 239061 from 0x60c000075650.
[DEBUG this=0x62900005a200] Erase 239062 from 0x60c000075650.
[DEBUG this=0x62900005a200] Erase 239063 from 0x60c000075650.
[DEBUG this=0x62900005a200] Erase 239064 from 0x60c000075650.
[DEBUG this=0x62900005a200] Erase 239065 from 0x60c000075650.
[DEBUG this=0x62900005a200] Erase 239066 from 0x60c000075650.
[DEBUG this=0x62900005a200] Erase 239067 from 0x60c000075650.
[DEBUG this=0x62900005a200; m_threadOrder.size() =  2, worker.GetThreadData().size() = 11] Insert threadData[ 2]=0x7fffccc36f96 in 0x60c000075650.
[DEBUG this=0x62900005a200; m_threadOrder.size() =  3, worker.GetThreadData().size() = 11] Insert threadData[ 3]=0x7fffccc47074 in 0x60c000075658.
[DEBUG this=0x62900005a200; m_threadOrder.size() =  4, worker.GetThreadData().size() = 11] Insert threadData[ 4]=0x7fffccc575cd in 0x60c000075660.
[DEBUG this=0x62900005a200; m_threadOrder.size() =  5, worker.GetThreadData().size() = 11] Insert threadData[ 5]=0x7fffccc67b26 in 0x60c000075668.
[DEBUG this=0x62900005a200; m_threadOrder.size() =  6, worker.GetThreadData().size() = 11] Insert threadData[ 6]=0x7fffccc784fa in 0x60c000075670.
[DEBUG this=0x62900005a200; m_threadOrder.size() =  7, worker.GetThreadData().size() = 11] Insert threadData[ 7]=0x7fffccc88a53 in 0x60c000075678.
[DEBUG this=0x62900005a200; m_threadOrder.size() =  8, worker.GetThreadData().size() = 11] Insert threadData[ 8]=0x7fffccc98b31 in 0x60c000075680.
[DEBUG this=0x62900005a200; m_threadOrder.size() =  9, worker.GetThreadData().size() = 11] Insert threadData[ 9]=0x7fffccca8cc3 in 0x60c000075688.
[DEBUG this=0x62900005a200; m_threadOrder.size() = 10, worker.GetThreadData().size() = 11] Insert threadData[10]=0x7fffcccbe3ad in 0x60c000075650.
Done sorting.
```

So, in conclusion, the bug is due to the fact the sorting code assumes any discrepancy is number of items in `m_threadOrder` and `worker.GetThreadData()` is _exclusively due to the elements in the end_ of `Worker::GetThreadData()`. On the surface, this is true, but the erasing logic to process _Thread Hints_ breaks this assumed invariant.

So to fix, in favor of performance, I propose to just clear the whole `m_threadOrder` vector when any pending thread hints are coming in. Those should happen rarely, I believe, which will cause a rebuild of the `m_threadOrder` vector.

I just tested this proposed fix, and confirm it works. Will open PR.

**Comment by wolfpld** [comment]
Thanks for looking into this. I'm not sure that the complete removal of the order of threads is a good thing from the user's point of view. While a new thread with a grouping hint will be rare, it can appear at any point in the capture, and by that time the user may have already ordered things to their liking.

**Comment by mcourteaux** [comment]
I see. Can you propose a fix? I can definitely test if you want to.

**Comment by mcourteaux** [comment]
> While a new thread with a grouping hint will be rare, it can appear at any point in the capture, and by that time the user may have already ordered things to their liking.

Doesn't `std::upper_bound` require a sorted list? If the user reorders threads, I don't know if the output of `std::upper_bound` is defined properly. I don't understand what the algorithm will do if it's not sorted, but partitioned:

> **Notes**
> Although std::upper_bound only requires [first, last) to be partitioned, this algorithm is usually used in the case where [first, last) is sorted, so that the binary search is valid for any value.

It does not specify what the algorithm does when the input is not partitioned. I think, if the user reorders, this could _technically, according to the spec_ be undefined behavior.

I think that a more exhaustive O(nk) insertion procedure is needed since the user can reorder (n=number of new elements, k=number of existing elements).

**Comment by mcourteaux** [comment]
Updated #879 with new fix!
